tempesta_fw/cfg.c

```diff
@@ -0,0 +1,1563 @@
+ * execution of this function, it may cause running modules to reference already
+ * stopped and un-initialized modules.To avoid that, we introduce the "cleanup"
+ * pass where modules may free memory after everything is stopped.
+ *
+ * Passes are done in reverse order of tfw_cfg_mod_start_all()
+ * (modules are started/stopped in LIFO manner).
+ */
+static void
+tfw_cfg_stop_mods(struct list_head *mod_list)
+{
+	TfwCfgMod *mod;
+
+	TFW_DBG("stopping modules...\n");
+
+	MOD_FOR_EACH_REVERSE(mod, mod_list) {
+		MOD_CALL(mod, stop);
+	}
+
+	MOD_FOR_EACH_REVERSE(mod, mod_list) {
+		MOD_CALL(mod, cleanup);
+	}
+
+}
+
+/*
+ * ------------------------------------------------------------------------
+ *	The list of registered modules, VFS and sysctl helpers.
+ * ------------------------------------------------------------------------
+ */
+
+/* The buffer net.tempesta.state value as a string.
+ * We need to store it to avoid double start or stop action.*/
+static char tfw_cfg_sysctl_state_buf[32];
+
+/* The file path is passed via the kernel module parameter.
+ * Usually you would not like to change it on a running system. */
+static char *tfw_cfg_path = "/etc/tempesta.conf";
+module_param(tfw_cfg_path, charp, 0444);
+MODULE_PARM_DESC(tfw_cfg_path, "Path to Tempesta FW configuration file.");
+
+/* The global list of all registered modules (consists of TfwCfgMod objects). */
+static LIST_HEAD(tfw_cfg_mods);
+
+/* The serialized value of tfw_cfg_sysctl_state_buf.
+ * Indicates that all registered modules are started. */
+bool tfw_cfg_mods_are_started;
+
+/**
+ * Read the whole file and put all the contents to the @out_buf.
+ */
+static int
+read_file_via_vfs(const char *path, char *out_buf, size_t buf_size)
+{
+	struct file *fp;
+	mm_segment_t oldfs;
+	loff_t offset = 0;
+	size_t bytes_read, read_size;
+	int ret = 0;
+
+	TFW_DBG("reading file: %s\n", path);
+
+	--buf_size; /* reserve one bye for '\0'. */
+	oldfs = get_fs();
+	set_fs(get_ds());
+
+	fp = filp_open(path, O_RDONLY, 0);
+	if (IS_ERR_OR_NULL(fp)) {
+		TFW_ERR("can't open file: %s (err: %ld)\n", path, PTR_ERR(fp));
+		goto out;
+	}
+
+	do {
+		buf_size -= offset;
+		read_size = min(buf_size, PAGE_SIZE);
+		bytes_read = vfs_read(fp, out_buf + offset, read_size, \
+				      &offset);
+		if (bytes_read < 0) {
+			ret = bytes_read;
+			TFW_ERR("can't read file: %s (err: %d)\n", path, ret);
+			goto out;
+		}
+	} while (bytes_read);
+
+out:
+	if (!IS_ERR_OR_NULL(fp))
+		filp_close(fp, NULL);
+	set_fs(oldfs);
+	out_buf[offset] = '\0';
+
+	return ret;
+}
+
+/**
+ * Process command received from sysctl as string (either "start" or "stop").
+ * Do corresponding actions, but only if the state is changed.
+ */
+static int
+handle_state_change(const char *old_state, const char *new_state)
+{
+	bool is_changed = strcasecmp(old_state, new_state);
+	bool is_start = !strcasecmp(new_state, "start");
+	bool is_stop = !strcasecmp(new_state, "stop");
```

Comment:
It seems new_state can't have "start" and "stop" line at the same time. However, you process is_start and is_stop variable as independent in the code. Rather, you should have single variable with ```start``` or ```stop``` value and process it w/o double checking in if statements below.